api/download.py

```diff
@@ -76,6 +76,52 @@ def archivestream(ticket):
 
 class Download(base.RequestHandler):
 
+    def _bulk_preflight_archivestream(self, file_refs):
+        data_path = config.get_item('persistent', 'data_path')
+        arc_prefix = 'sdm'
+        file_cnt = 0
+        total_size = 0
+        targets = []
+
+        for fref in file_refs:
+            cont_name   = fref.get('container_name','')+'s'
+            cont_id     = fref.get('container_id', '')
+            filename    = fref.get('filename', '')
+
+            if cont_name not in ['projects', 'sessions', 'acquisitions']:
+                self.abort(400, 'Bulk download only supports files in projects, sessions and acquisitions')
+            file_obj = None
+            try:
+                # Try to find the file reference in the database (filtering on user permissions)
+                bid = bson.ObjectId(cont_id)
+                file_obj = config.db[cont_name].find_one(
+                    {'_id': bid, 'permissions._id': self.uid},
+                    {'files': { '$elemMatch': {
+                        'name': filename
+                    }}
+                })['files'][0]
+            except:
+                # self.abort(404, 'File {} on Container {} {} not found'.format(filename, cont_name, cont_id))
+                # silently skip missing files/files user does not have access to
+                continue
+
+            filepath = os.path.join(data_path, util.path_from_hash(file_obj['hash']))
+            if os.path.exists(filepath): # silently skip missing files
+                targets.append((filepath, cont_id+'-'+file_obj['name'], file_obj['size']))
```

Comment:
Currently we're naming the file `<cont_id>-<filename>` for uniqueness which is not a great solution. Let me know if you have any other ideas. Below we create the hierarchy in a file system. For this I wanted something a bit simpler and quicker :).